src/core/src/pattern/op/optional.cpp

```diff
@@ -0,0 +1,49 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "openvino/pass/pattern/op/optional.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+#include "openvino/pass/pattern/matcher.hpp"
+
+std::unordered_set<ov::DiscreteTypeInfo> ov::pass::pattern::op::Optional::get_optional_types() const {
+    return optional_types;
+}
+
+bool ov::pass::pattern::op::Optional::match_value(Matcher* matcher,
+                                                  const Output<Node>& pattern_value,
+                                                  const Output<Node>& graph_value) {
+    const auto p_value = input_value(0);
+    const auto p_node = p_value.get_node_shared_ptr();
+    auto& pattern_map = matcher->get_pattern_value_map();
```

Comment:
Can we use pattern::Or<parent_node_in_the_pattern_graph, WrapType(optional_types)> here as we discussed yesterday?  I believe it would be more solid solution with the same complexity.

Reply:
@itikhono 
As I mentioned in conversiations, I don't see any reasons to change the optional logic to use `OR` because of:
* source code won't be reduced a lot according collection of patterns for OR operation
* Extra call for OR comparation as an extra step for call stack